### ISSUE TITLE
[REVIEW]: Easy SimAuto (ESA): A Python Package that Simplifies Interacting with PowerWorld Simulator

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2289]
**Submitting author:** @blthayer (<a href="http://orcid.org/0000-0002-6517-1295">Brandon Thayer</a>)
**Repository:** <a href="https://github.com/mzy2240/ESA" target ="_blank">https://github.com/mzy2240/ESA</a>
**Version:** v1.0.6
**Editor:** @danielskatz
**Reviewer:** @robinroche, @tonyelhabr
**Archive:** <a href="https://doi.org/10.5281/zenodo.3899462" target="_blank">10.5281/zenodo.3899462</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37/status.svg)](https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37"><img src="https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37/status.svg)](https://joss.theoj.org/papers/e86f2f9186fcfefa6b4a53e1c35f3f37)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/mzy2240/ESA" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@robinroche & @tonyelhabr, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mzy2240/ESA">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@blthayer) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mzy2240/ESA">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@blthayer) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @robinroche, @tonyelhabr it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/DEXA.2012.9 is OK
- 10.1109/TPEC48276.2020.9042493 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02289/joss.02289/10.21105.joss.02289.pdf)

**Comment by danielskatz** [comment]
👋 @robinroche & @tonyelhabr -  thanks for agreeing to review this submission. Please be sure to read the comments above, and let me know if you have any questions. Basically, your job is to check the article proof and repository and check items off your checklist above. (And note that you will have to accept the invitation before you will be able to do so.)

If you see small problems that need to be discussed, feel free to discuss them here. But if you can, create a new issue in the target repository and link to this review thread in that issue to create a corresponding breadcrumb trail here.

I look forward to seeing how this review goes

**Comment by blthayer** [comment]
@danielskatz - Thank you for kicking this off! I don't see `tonyelhabr` listed as an assignee on this issue - is that expected?

Also, I wanted to note that we tweaked the LaTex template so that a footnote appears indicating that myself and @mzy2240 contributed equally to the software. However, that footnote isn't showing up in the automatically generated .pdf file. What steps do we need to take to use our customized template before final publication? This isn't urgent since the review itself is still pending, but I figured it's best to bring it up earlier rather than later.

**Comment by danielskatz** [comment]
> Thank you for kicking this off! I don't see tonyelhabr listed as an assignee on this issue - is that expected?

I think he will be marked as assigned once he accepts the invite as mentioned in the first comment in this issue.

>  What steps do we need to take to use our customized template before final publication?

@arfon - can you help here?

**Comment by arfon** [comment]
Unfortunately that kind of custom modification isn't supported @blthayer. That said, something similar is possible using our existing template. I've downloaded your paper locally and changed the authors block to:

```
authors:
  - name: Brandon L. Thayer^[The first two authors of this paper contributed equally to this software.]
    orcid: 0000-0002-6517-1295
    affiliation: "1, 2"
  - name: Zeyu Mao
    orcid: 0000-0003-0841-5123
    affiliation: 1
  - name: Yijing Liu
    orcid: 0000-0002-5104-325X
    affiliation: 1
  - name: Katherine Davis
    orcid: 0000-0002-1603-1122
    affiliation: 1
  - name: Thomas Overbye
    orcid: 0000-0002-2382-2811
    affiliation: 1
```

...which results in the following PDF: [10.21105.joss.02289.pdf](https://github.com/openjournals/joss-reviews/files/4764477/10.21105.joss.02289.pdf)

**Comment by blthayer** [comment]
@danielskatz - thank you for the clarification, and for bringing @arfon into the discussion.

@arfon - thank you for the tip! I'll get things tweaked shortly.

**Comment by tonyelhabr** [comment]
@whedon My invitation expired. Can you send another?

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by danielskatz** [comment]
@whedon re-invited @tonyelhabr as reviewer

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by tonyelhabr** [comment]
@whedon re-invite @tonyelhabr  as reviewer

**Comment by whedon** [comment]
I'm sorry @tonyelhabr, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by danielskatz** [comment]
@whedon re-invite @tonyelhabr as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@tonyelhabr please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
👋 @tonyelhabr  - I see you've checked off all the items in your list, which I assume means you didn't find anything that needed to be improved - the submission meets all the required criteria.  Do you have any comments on the paper or software otherwise?

**Comment by blthayer** [comment]
@danielskatz - I wanted to check if it's acceptable at this stage to swap out a single reference. Specifically, reference `(Thayer, 2020c)`. That reference refers to my Master's thesis, but the full text will not be publicly available until the Spring of 2022 (delayed release is conventional practice to avoid copyright issues when publishing portions of the thesis in academic journals). I've condensed my thesis into a journal article submission, and that submission should be available on arXiv starting Monday. So, I'd like to reference the arXiv journal article instead of the original thesis (which is currently unavailable except by request).

**Comment by danielskatz** [comment]
Until this is published, any changes are up to you, though you should let the reviewers know what you are changing and why.

**Comment by tonyelhabr** [comment]
@danielskatz I have no suggestions. I tested all examples in the documentation to verify software functionality. The examples are all non-trivial (with the exception of "Quick Start") and do bear resemblance to real-world usage (from my experience), which is good to see. The paper is concise and meets all of checklist items.

**Comment by blthayer** [comment]
@danielskatz - thank you!
@tonyelhabr - thanks again for your review!

@tonyelhabr, @robinroche - please see [this comment](https://github.com/openjournals/joss-reviews/issues/2289#issuecomment-643388272) with regards to a very minor tweak I plan to make to a reference contained in the paper. I plan to make the reference change early next week (pending approval from arXiv).

I hope you all have great weekends!

**Comment by robinroche** [comment]
@blthayer 
I installed ESA without any issue and I am currently going through the quick start steps.
I get the following error:

```
(tutorial-env) C:\Users\rroche\Downloads>python test.py
22:55:06 [WARNING] [SAW]: UIVisible attribute could not be set. Note this SimAuto property was not introduced until Simulator version 20. Check your version with the get_simulator_version method.
Traceback (most recent call last):
  File "pandas\_libs\lib.pyx", line 1926, in pandas._libs.lib.maybe_convert_numeric
ValueError: Unable to parse string "  1,05999994"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 8, in <module>
    bus_data = saw.get_power_flow_results('bus')
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 458, in get_power_flow_results
    return self.GetParametersMultipleElement(ObjectType=object_type,
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 904, in GetParametersMultipleElement
    return self.clean_df_or_series(obj=df, ObjectType=ObjectType)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 284, in clean_df_or_series
    obj[numeric_fields] = obj[numeric_fields].apply(pd.to_numeric)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\frame.py", line 6878, in apply
    return op.get_result()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 186, in get_result
    return self.apply_standard()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 313, in apply_standard
    results, res_index = self.apply_series_generator()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 341, in apply_series_generator
    results[i] = self.f(v)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\tools\numeric.py", line 149, in to_numeric
    values = lib.maybe_convert_numeric(
  File "pandas\_libs\lib.pyx", line 1963, in pandas._libs.lib.maybe_convert_numeric
ValueError: Unable to parse string "  1,05999994" at position 0
```
The problem appears to be with the last line of:
```
CASE_PATH = r"C:\Users\rroche\Downloads\IEEE 14 bus_pws_version_17.pwb"

from esa import SAW

saw = SAW(FileName=CASE_PATH)
saw.SolvePowerFlow()

bus_data = saw.get_power_flow_results('bus')
```
The warning seems normal as I am using PWS17.
For the parsing error, I suspect (but I may be wrong) that it may be related to the fact that I am using Windows 10 in French, for which the decimal point is actually a coma.
Running the same script without the last line returns no error, just the warning.

**Comment by blthayer** [comment]
@robinroche - thank you for bringing up this issue!

I agree - the warning you received is to be expected.

I have added support for comma-delimited decimal numbers. If you would like to get your hands on the changes right away, you can install ESA from source from our `develop` branch (found [here](https://github.com/mzy2240/ESA/tree/develop)).

Alternatively, I have submitted a pull request to be reviewed by the other ESA maintainers. After they've reviewed, a new ESA release will published including this fix. Regardless of which method you prefer, I will comment here when the next version is published and available to install via pip.

Thanks again, and have a great weekend!

**Comment by robinroche** [comment]
@blthayer thanks, I'll try the new version as soon as it is released then.

**Comment by blthayer** [comment]
@robinroche - the aforementioned changes have been released in ESA version `1.0.4`. I have noted the updates in the [changelog](https://mzy2240.github.io/ESA/html/welcome.html#changelog).

Assuming you installed ESA via pip, the process to upgrade is the same that you used for your initial installation, but you'll need to use pip's `--upgrade` flag. For example:

`python -m pip install --upgrade esa`

You can check the ESA version afterwards via:

```python
import esa
print(esa.__version__)
```

As I mentioned in my previous comment, please let me know if this fix doesn't work for you by commenting on [this issue](https://github.com/mzy2240/ESA/issues/56#issue-638041265). If the fix doesn't work for you, I'll work with you to collect more details so I can more directly reproduce the problem on my side and completely fix it.

**Comment by blthayer** [comment]
@robinroche - we have a fix coming your way in ESA version `1.0.5`. I'll comment here when that's officially released.

If you would prefer to get started right away, adding the following to the _very top_ of your script should fix the issue for version `1.0.4`:

```python
import locale
locale.setlocale(locale.LC_ALL, '')
```

**Comment by robinroche** [comment]
@blthayer It worked, thanks. I continue the tests.

**Comment by blthayer** [comment]
@robinroche - excellent, thank you! 

Also note that version `1.0.5` is now available if you'd like to upgrade. There's a very small bug fix in `1.0.5` in addition to adding the aforementioned `locale` lines to `saw.py`, but I don't think the bug affects any of the examples.

**Comment by robinroche** [comment]
@blthayer Thanks, I upgraded to the latest version.
I'm now testing the "Add Lines to Case" example.
I'm getting the following:
```
(tutorial-env) C:\Users\rroche\Downloads>python test5.py
   From Number  To Number  Ckt        R        X        B  Lim MVA A
0         8155       5358    3  0.00037  0.00750  0.52342       2768
1         8154       8135    3  0.00895  0.03991  0.00585        149
2         8153       8108    3  0.01300  0.05400  0.02700        186
3         8152       8160    3  0.00538  0.03751  0.00613        221
4         8155       8057    3  0.00037  0.00750  0.52342       2768
5         8154       8153    3  0.01300  0.05400  0.02700        186
6         8155       8135    3  0.00538  0.03751  0.00613        221
19:18:30 [WARNING] [SAW]: UIVisible attribute could not be set. Note this SimAuto property was not introduced until Simulator version 20. Check your version with the get_simulator_version method.
Index(['BusNum', 'BusNum:1', 'LineCircuit', 'LineR', 'LineX', 'LineC',
       'LineAMVA'],
      dtype='object')
['BusNum', 'BusNum:1', 'LineCircuit']
Traceback (most recent call last):
  File "test5.py", line 34, in <module>
    saw.change_and_confirm_params_multiple_element('branch', line_df)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 202, in change_and_confirm_params_multiple_element
    cleaned_df = self._change_parameters_multiple_element_df(
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1754, in _change_parameters_multiple_element_df
    cleaned_df = self.clean_df_or_series(obj=command_df,
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 290, in clean_df_or_series
    obj[numeric_fields] = self._to_numeric(obj[numeric_fields])
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1852, in _to_numeric
    data = data.apply(lambda x:
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\frame.py", line 6878, in apply
    return op.get_result()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 186, in get_result
    return self.apply_standard()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 295, in apply_standard
    result = libreduction.compute_reduction(
  File "pandas\_libs\reduction.pyx", line 620, in pandas._libs.reduction.compute_reduction
  File "pandas\_libs\reduction.pyx", line 128, in pandas._libs.reduction.Reducer.get_result
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1853, in <lambda>
    x.str.replace(self.decimal_delimiter, '.'))
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\generic.py", line 5270, in __getattr__
    return object.__getattribute__(self, name)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\accessor.py", line 187, in __get__
    accessor_obj = self._accessor(obj)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2041, in __init__
    self._inferred_dtype = self._validate(data)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2098, in _validate
    raise AttributeError("Can only use .str accessor with string values!")
AttributeError: Can only use .str accessor with string values!
```
Lines 33-34 in the script are:
```
(33) saw.RunScriptCommand("EnterMode(EDIT);")
(34) saw.change_and_confirm_params_multiple_element('branch', line_df)
```
I don't know if this is another locale issue, so just in case, I tried adding the following to the script, but the result remains the same:
```
import locale
locale.setlocale(locale.LC_ALL, '')
```

**Comment by blthayer** [comment]
@robinroche - thank you! I'll file an issue in the ESA repository, and work on this issue this evening (Pacific time, West coast of USA). 

I do suspect this is at least tangentially related to the locale issue, as I ran this example on my local machine and it works. So, I'll likely be giving you commands to execute so that I can get your output and reproduce on my end (like we did last time).

Thanks again for your time and review - your help in making our software work for older versions of PowerWorld and countries outside the US is **greatly** appreciated. 

In the mean time, if you're able/willing to skip around to other examples so we can get any other issues knocked out as quickly as possible, I would appreciate it. Please note that one of the examples (the [weighted graph example](https://mzy2240.github.io/ESA/html/snippets.html#created-graph-model-with-edges-weighted-by-branch-impedance)) does not work with PWS 17, and the fact that it will not work with PWS 17 is noted in the docs. 

Thanks!

**Comment by robinroche** [comment]
@blthayer So I tried all examples (see below). The good news is that only 2 examples are not working in my case, and the problem seems to be the same for both.

 - Install: OK
 - Quick start: OK (with 1.0.5)
 - Increase loading in case: error #1
 - Add lines to case: error #2
 - Create simple graph model: OK
 - Create graph model with edges weighted bu branch impedance: unable to test due to PWS17
 - Plot histogram of line flows with matplotlib: OK
 
 
 ERROR #1
 ```
 Traceback (most recent call last):
  File "test_increaseloading.py", line 16, in <module>
    saw.change_parameters_multiple_element_df('load', load_frame)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 246, in change_parameters_multiple_element_df
    self._change_parameters_multiple_element_df(
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1754, in _change_parameters_multiple_element_df
    cleaned_df = self.clean_df_or_series(obj=command_df,
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 290, in clean_df_or_series
    obj[numeric_fields] = self._to_numeric(obj[numeric_fields])
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1852, in _to_numeric
    data = data.apply(lambda x:
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\frame.py", line 6878, in apply
    return op.get_result()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 186, in get_result
    return self.apply_standard()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 295, in apply_standard
    result = libreduction.compute_reduction(
  File "pandas\_libs\reduction.pyx", line 620, in pandas._libs.reduction.compute_reduction
  File "pandas\_libs\reduction.pyx", line 128, in pandas._libs.reduction.Reducer.get_result
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1853, in <lambda>
    x.str.replace(self.decimal_delimiter, '.'))
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\generic.py", line 5270, in __getattr__
    return object.__getattribute__(self, name)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\accessor.py", line 187, in __get__
    accessor_obj = self._accessor(obj)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2041, in __init__
    self._inferred_dtype = self._validate(data)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2098, in _validate
    raise AttributeError("Can only use .str accessor with string values!")
AttributeError: Can only use .str accessor with string values!
```

ERROR #2
```
Traceback (most recent call last):
  File "test_addlines.py", line 34, in <module>
    saw.change_and_confirm_params_multiple_element('branch', line_df)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 202, in change_and_confirm_params_multiple_element
    cleaned_df = self._change_parameters_multiple_element_df(
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1754, in _change_parameters_multiple_element_df
    cleaned_df = self.clean_df_or_series(obj=command_df,
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 290, in clean_df_or_series
    obj[numeric_fields] = self._to_numeric(obj[numeric_fields])
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1852, in _to_numeric
    data = data.apply(lambda x:
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\frame.py", line 6878, in apply
    return op.get_result()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 186, in get_result
    return self.apply_standard()
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\apply.py", line 295, in apply_standard
    result = libreduction.compute_reduction(
  File "pandas\_libs\reduction.pyx", line 620, in pandas._libs.reduction.compute_reduction
  File "pandas\_libs\reduction.pyx", line 128, in pandas._libs.reduction.Reducer.get_result
  File "C:\Users\rroche\tutorial-env\lib\site-packages\esa\saw.py", line 1853, in <lambda>
    x.str.replace(self.decimal_delimiter, '.'))
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\generic.py", line 5270, in __getattr__
    return object.__getattribute__(self, name)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\accessor.py", line 187, in __get__
    accessor_obj = self._accessor(obj)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2041, in __init__
    self._inferred_dtype = self._validate(data)
  File "C:\Users\rroche\tutorial-env\lib\site-packages\pandas\core\strings.py", line 2098, in _validate
    raise AttributeError("Can only use .str accessor with string values!")
AttributeError: Can only use .str accessor with string values!
```

**Comment by blthayer** [comment]
@robinroche - thanks for your update, and thanks for the details.

We have a new release that _should_ address these issues: `v1.0.6`. If you could upgrade ESA and try those examples again, I'm hopeful that they'll work for you now.

Thanks again for your time and willingness to volunteer and help.

**Comment by robinroche** [comment]
@blthayer I upgraded to 1.0.6 and all examples now work, thanks!

 - Install: OK
 - Quick start: OK
 - Increase loading in case: OK
 - Add lines to case: OK
 - Create simple graph model: OK
 - Create graph model with edges weighted by branch impedance: unable to test due to PWS17
 - Plot histogram of line flows with matplotlib: OK

**Comment by blthayer** [comment]
@robinroche - excellent, I'm glad to hear it! Thanks for bearing with us as we addressed those locale issues. 

As you go through the JOSS checklist, we'd love to hear further suggestions for improving ESA. Please don't hesitate to file issues over in our repository, even if it's as simple as "Section X of the documentation is confusing and should be revised."

**Comment by robinroche** [comment]
@blthayer I have finished the review. Congratulations on this software, I wish it had existed when I needed it during my PhD! It is a much needed addition to PWS/SimAuto, simple to use and well documented.

Here are some minor improvement suggestions:
- In the documentation landing page, briefly explain what is SimAuto and what it can do (e.g., a few simple examples).
- Make a link to SimAuto's documentation more visible (I would expect most users would need it at some point).
- Clarify that the case path is the path to the .pwb file.
- State the goal of each example, and show results whenever possible (e.g., with power flow, but I acknowledge this may be difficult to keep it simple).
- List the main functions of SimAuto, so people can have an idea of what is possible with it and with ESA (with a link to SimAuto's documentation).

@danielskatz I have finished the review.

**Comment by danielskatz** [comment]
Thanks @robinroche!
@blthayer - please go ahead and make the changes suggested by @robinroche, or let us know what you disagree with.
Once this is done, we can move forward on the acceptance of the submission.

**Comment by blthayer** [comment]
@robinroche - thank you! I appreciate your suggestions, and we'll get those worked in shortly.

@danielskatz, @mzy2240 - I'll plan to address @robinroche's comments tomorrow, either in the morning before work or in the evening afterwards.

**Comment by blthayer** [comment]
@robinroche, @danielskatz: I have addressed the documentation suggestions. 

- Added new [About SimAuto](https://mzy2240.github.io/ESA/html/welcome.html#about-simauto) page and added text with links to the [table of contents/landing page](https://mzy2240.github.io/ESA/html/index.html)
- Links to SimAuto's documentation are available via the previously mentioned links
- All [examples](https://mzy2240.github.io/ESA/html/snippets.html) have language like `Before running the example below, define a CASE_PATH constant (the file path to a PowerWorld .pwb case file)` for clarification
- Any examples which did not already have their purpose stated up front now do
- All SimAuto functions and properties are listed in [About SimAuto](https://mzy2240.github.io/ESA/html/welcome.html#about-simauto)

**Comment by danielskatz** [comment]
Thanks - @robinroche, please let us know how this seems to you when you get a chance.

**Comment by robinroche** [comment]
@blthayer @danielskatz All suggestions have been addressed. This matches what I had in mind, thanks.

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by danielskatz** [comment]
@whedon check references

**Comment by whedon** [comment]
PDF failed to compile for issue #2289 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/lib/whedon/compilers.rb:158:in `initialize': No such file or directory @ rb_sysopen - tmp/2289/docs/joss/markdown-metadata.yaml (Errno::ENOENT)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/lib/whedon/compilers.rb:158:in `open'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/lib/whedon/compilers.rb:158:in `pdf_from_markdown'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/lib/whedon/compilers.rb:13:in `generate_pdf'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/bin/whedon:56:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-2081a6a6226c/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/DEXA.2012.9 is OK
- 10.1109/TPEC48276.2020.9042493 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02289/joss.02289/10.21105.joss.02289.pdf)

**Comment by blthayer** [comment]
@danielskatz - the pdf looks great to me! I didn't find any issues.

**Comment by danielskatz** [comment]
Can you merge the changes in https://github.com/mzy2240/ESA/pull/63 or let me know what you disagree with?

**Comment by danielskatz** [comment]
Then the next steps are for you to:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g. figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by blthayer** [comment]
@danielskatz:

- tag; `v1.0.6
- Zenodo DOI: `10.5281/zenodo.3899462`

**Comment by danielskatz** [comment]
@whedon set v1.0.6 as version

**Comment by whedon** [comment]
OK. v1.0.6 is the version.

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.3899462 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3899462" target="_blank">10.5281/zenodo.3899462</a> is the archive.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/DEXA.2012.9 is OK
- 10.1109/TPEC48276.2020.9042493 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1493

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1493, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1273345478088892416) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1494
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02289](https://doi.org/10.21105/joss.02289)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @robinroche & @tonyelhabr for reviewing!

And congratulations to @blthayer (Brandon Thayer) and co-authors!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02289/status.svg)](https://doi.org/10.21105/joss.02289)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02289">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02289/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02289/status.svg
   :target: https://doi.org/10.21105/joss.02289
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02289/status.svg)](https://doi.org/10.21105/joss.02289)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by blthayer** [comment]
@danielskatz - thank you for your time and effort! Publishing in JOSS was a great experience, and I'll be paying it forward by signing up as a reviewer.

**Comment by al3xa23** [comment]
Hi @blthayer, the editorialbot suggest you as reviewer. Is that with you?
